globus_cli/parsing/shell_completion.py

```diff
@@ -0,0 +1,195 @@
+        if not args:
+            break
+
+        # check for a matching command, and if one isn't found stop the
+        # traversal and abort the whole process -- this would mean that a
+        # completed command was entered which doesn't match a known command
+        # there's nothing completion can do in this case unless it implements
+        # sophisticated fuzzy matching
+        command = ctx.command.get_command(ctx, args[0])
+        if not command:
+            return None
+
+        # otherwise, grab that command, and build a subcontext to continue the
+        # tree walk
+        else:
+            ctx = command.make_context(args[0], args[1:], parent=ctx,
+                                       resilient_parsing=True)
+
+    # return the context we found
+    return ctx
+
+
+def get_all_choices(completed_args, cur, quoted):
+    ctx = get_completion_context(completed_args)
+    if not ctx:
+        return []
+
+    # matching rules, so we can toggle by type and such
+    def match_with_case(n, m):
+        """case matters"""
+        return n.startswith(m)
+
+    def match_nocase(n, m):
+        """case doesn't matter"""
+        return n.lower().startswith(m.lower())
+
+    match_func = match_with_case
+
+    ctx_options = [p for p in ctx.command.get_params(ctx)
+                   if isinstance(p, click.Option)]
+
+    last_completed = None
+    if completed_args:
+        last_completed = completed_args[-1]
+
+    # if the last completed argument matches a Choice option, we're going to
+    # have to expand cur as a choice param
+    matching_choice_opt = None
+    for p in ctx_options:
+        if isinstance(p.type, click.Choice) and last_completed in p.opts:
+            matching_choice_opt = p
+
+    choices = []
+    # if we ended on a choice, complete with all of the available values
+    if matching_choice_opt:
+        # catch the case where it's case insensitive, and we need to change our
+        # comparisons / matching later on
+        if isinstance(matching_choice_opt.type, CaseInsensitiveChoice):
+            match_func = match_nocase
+        choices = matching_choice_opt.type.choices
+    # if cur looks like an option, just look for options
+    # but skip if it's quoted text
+    elif cur and cur.startswith('-') and not quoted:
+        for param in ctx_options:
+            # skip hidden options
+            if isinstance(param, HiddenOption):
+                continue
+            for opt in param.opts:
+                # only add long-opts, never short opts to completion, unless
+                # the cur appears to be a short opt already
+                if opt.startswith('--') or (
+                        len(cur) > 1 and cur[1] != '-'):
+                    choices.append(opt)
+    # and if it's a multicommand we see, get the list of subcommands
+    elif isinstance(ctx.command, click.MultiCommand) and not quoted:
+        choices = [name for name in ctx.command.list_commands(ctx)]
```

Comment:
This list comprehension doesn't seem to be necessary.
Maybe you used to have a predicate here or maybe you just want a shallow copy (`list()` or `[:]`?).